File: 프로그래머스/2/169199. 리코쳇 로봇/리코쳇 로봇.py
from collections import deque

def solution(board):
    answer = 0
    
    start_x, start_y = 0, 0
    
    n, m = len(board), len(board[0])
    
    start = False
    for a in range(n):
        if start:
            break
        for b in range(m):
            if board[a][b] == 'R':
                start_x, start_y = a, b
                start = True
                break
    
    queue = deque([(start_x, start_y, 0)]) # 좌표, 이동 횟수
    visited = set()
    visited.add((start_x, start_y))
    
    directions = [(-1, 0), (1, 0), (0, -1), (0, 1)]
    
    while queue:
        x, y, moves = queue.popleft()
        
        if board[x][y] == 'G':
            return moves
        
        for dx, dy in directions:
            nx, ny = x, y
            
            while 0 <= nx + dx < n and 0 <= ny + dy < m and board[nx + dx][ny + dy] != 'D':
                nx += dx
                ny += dy
            if (nx, ny) not in visited:
                visited.add((nx, ny))
                queue.append((nx, ny, moves + 1))
                
    return -1

File: 프로그래머스/2/169199. 리코쳇 로봇/리코쳇 로봇.java
import java.util.*;

class Solution {
    static class Node{
        int x, y, count;
        Node(int x, int y, int count) {
            this.x = x;
            this.y = y;
            this.count = count;
        }
    }
    
    public int solution(String[] board) {
        int n = board.length;
        int m = board[0].length();
        char[][] map = new char[n][m];
        int sx = 0, sy = 0, gx = 0, gy = 0;
        
        for (int i = 0; i < n; i++) {
            map[i] = board[i].toCharArray();
            for (int j = 0; j < m; j++) {
                if (map[i][j] == 'R') {
                    sx = i;
                    sy = j;
                } else if (map[i][j] == 'G') {
                    gx = i;
                    gy = j;
                }
            }
        }
        
        Queue<Node> q = new LinkedList<>();
        boolean[][] visited = new boolean[n][m];
        q.add(new Node(sx, sy, 0));
        visited[sx][sy] = true;
        int[] dx = {-1, 1, 0, 0};
        int[] dy = {0, 0, -1, 1};
        
        while (!q.isEmpty()) {
            Node current = q.poll();
            if (current.x == gx && current.y == gy) {
                return current.count;
            }
            
            for (int d = 0; d < 4; d++) {
                int x = current.x;
                int y = current.y;
                while(true) {
                    int nx = x + dx[d];
                    int ny = y + dy[d];
                    if (nx < 0 || nx >= n || ny < 0 || ny >= m) break;
                    if (map[nx][ny] == 'D') break;
                    x = nx;
                    y = ny;
                }
                if (!visited[x][y]) {
                    visited[x][y] = true;
                    q.add(new Node(x, y, current.count + 1));
                }
            }
        }
        
        return -1;
    }
}